app/src/main/java/org/schabi/newpipe/player/mediasession/PlayQueueNavigator.java

```diff
@@ -1,106 +1,152 @@
     }
 
     @Override
-    public void onSkipToPrevious(@NonNull final Player player) {
-        callback.playPrevious();
+    public void onSkipToPrevious(@NonNull final com.google.android.exoplayer2.Player exoPlayer) {
+        player.playPrevious();
     }
 
     @Override
-    public void onSkipToQueueItem(@NonNull final Player player, final long id) {
-        callback.playItemAtIndex((int) id);
+    public void onSkipToQueueItem(@NonNull final com.google.android.exoplayer2.Player exoPlayer,
+                                  final long id) {
+        if (player.getPlayQueue() != null) {
+            player.selectQueueItem(player.getPlayQueue().getItem((int) id));
+        }
     }
 
     @Override
-    public void onSkipToNext(@NonNull final Player player) {
-        callback.playNext();
+    public void onSkipToNext(@NonNull final com.google.android.exoplayer2.Player exoPlayer) {
+        player.playNext();
     }
 
     private void publishFloatingQueueWindow() {
-        if (callback.getQueueSize() == 0) {
+        final int windowCount = Optional.ofNullable(player.getPlayQueue())
+                .map(PlayQueue::size)
+                .orElse(0);
+        if (windowCount == 0) {
             mediaSession.setQueue(Collections.emptyList());
             activeQueueItemId = MediaSessionCompat.QueueItem.UNKNOWN_ID;
             return;
         }
 
         // Yes this is almost a copypasta, got a problem with that? =\
-        final int windowCount = callback.getQueueSize();
-        final int currentWindowIndex = callback.getCurrentPlayingIndex();
-        final int queueSize = Math.min(maxQueueSize, windowCount);
+        final int currentWindowIndex = player.getPlayQueue().getIndex();
+        final int queueSize = Math.min(MAX_QUEUE_SIZE, windowCount);
         final int startIndex = Util.constrainValue(currentWindowIndex - ((queueSize - 1) / 2), 0,
                 windowCount - queueSize);
 
         final List<MediaSessionCompat.QueueItem> queue = new ArrayList<>();
         for (int i = startIndex; i < startIndex + queueSize; i++) {
-            queue.add(new MediaSessionCompat.QueueItem(callback.getQueueMetadata(i), i));
+            queue.add(new MediaSessionCompat.QueueItem(getQueueMetadata(i), i));
         }
         mediaSession.setQueue(queue);
         activeQueueItemId = currentWindowIndex;
     }
 
+    public MediaDescriptionCompat getQueueMetadata(final int index) {
+        if (player.getPlayQueue() == null) {
+            return null;
+        }
+        final PlayQueueItem item = player.getPlayQueue().getItem(index);
+        if (item == null) {
+            return null;
+        }
+
+        final MediaDescriptionCompat.Builder descBuilder = new MediaDescriptionCompat.Builder()
+                .setMediaId(String.valueOf(index))
+                .setTitle(item.getTitle())
+                .setSubtitle(item.getUploader());
+
+        // set additional metadata for A2DP/AVRCP
```

Comment:
What's ``A2DP/AVRCP``?

Reply:
I thought you would have Wiki'd it immediately. https://en.wikipedia.org/wiki/List_of_Bluetooth_profiles

Reply:
I have no idea, I didn't write this code, I just copied it from [here](https://github.com/TeamNewPipe/NewPipe/blob/cacce6d2d0946ecf2b3cd79e20129031c2c0ca45/app/src/main/java/org/schabi/newpipe/player/playback/PlayerMediaSession.java#L69) ;-)
I will improve the comment anyway.

Reply:
Hmm okay, maybe we should fine tune the comment in the future to something like "set additional metadata for Bluetooth" - save a Wikipedia page call :D